source/fundamentals/csfle.txt

```diff
@@ -0,0 +1,224 @@
+
+``mongocryptd``
+~~~~~~~~~~~~~~~
+
+``mongocryptd`` is a binary run as a daemon / process that is used for 
+automatic encryption. ``mongodb-crypt`` communicates with 
+``mongodb-client-encryption`` to automatically encrypt the information 
+specified by a user-provided 
+:manual:`JSON Schema </reference/security-client-side-automatic-json-schema/>`.
+
+For more detailed information on ``mongocryptd``, see the
+:manual:`mongocryptd reference documentation </reference/security-client-side-encryption-appendix/#mongocryptd>`
+
+Example
+-------
+
+The following example shows how to configure a CSFLE-enabled client 
+with a local key and a JSON schema. Values in the ``ssn`` field are 
+automatically encrypted before insertion, and decrypted when calling 
+``find()`` with a CSFLE-enabled client.
+
+.. warning::
+
+   MongoDB recommends using local key management only for testing 
+   purposes, and using a remote key management service
+   for production.
+
+An expanded example with support for remote key management services is 
+available at MongoDB University's GitHub 
+`Node CSFLE Example <https://github.com/mongodb-university/csfle-guides/tree/master/nodejs>`__.
+
+.. note::
+   
+   Auto encryption requires MongoDB **Enterprise** or **Atlas**.
+
+To run this example, first complete the following steps:
+
+- Save `master-key.txt <https://github.com/mongodb-university/csfle-guides/raw/master/nodejs/master-key.txt>`__
+  to the same directory as this example code.
+- Start a ``mongod`` locally on the default port 27017.
+- Start a ``mongocryptd`` locally on the default port 27020.
+
+.. code-block:: javascript
+  
+   const { MongoClient, Binary } = require("mongodb");
+   const { ClientEncryption } = require("mongodb-client-encryption");
+   const fs = require("fs");
+   
+   async function getRegularClient() {
+     const client = new MongoClient("mongodb://localhost:27017", { 
+       useNewUrlParser: true,
+       useUnifiedTopology: true,
+     });
+     return await client.connect();
+   }
+   
+   async function getCsfleEnabledClient(schemaMap) {
+     const client = new MongoClient("mongodb://localhost:27017", {
+       useNewUrlParser: true,
+       useUnifiedTopology: true,
+       autoEncryption: {
+         keyVaultNamespace: "encryption.__keyVault",
+         kmsProviders: {
+           local: {
+             key: fs.readFileSync("./master-key.txt"),
+           },
+         },
+         schemaMap,
+       },
+     });
+     return await client.connect();
+   }
+   
+   function createJsonSchemaMap(dataKey) {
+     return {
+       "users.ssns": {
+         bsonType: "object",
+         encryptMetadata: {
+           keyId: [new Binary(Buffer.from(dataKey, "base64"), 4)],
+         },
+         properties: {
+           ssn: {
+             encrypt: {
+               bsonType: "int",
+               algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic",
+             },
+           },
```

Comment:
Total suggestion here:
We can leave this as is, it doesn't change the usefulness of the example at all, but to my knowledge its common practice to store something like a social security number as a string, it doesn't have as much useful meaning as an int. For example, it has a certain length that is required, and you often want to check the last four digits. Neither of those operations are particularly easy with an Int but are with a string.

Reply:
Got it, that makes a lot of sense. Still, I'm planning to leave this as-is to match the University example as closely as possible.